Transpile testing

abidlabs opened the pull request.

abidlabs commented on js/core/src/Blocks.svelte:
```diff
@@ -391,6 +406,9 @@
 			submit_map.set(dep_index, submission);
 
 			for await (const message of submission) {
+				if (payload.js_implementation) {
```
If a function has a js_implementation, we still want to make the API call to the backend since it might have downstream effects. For example, updating the "type" of an "Image" could affect the processing of a subsequent call.

freddyaboulton replied:
So it runs both on the frontend and backend?

abidlabs replied:
Yes, maybe we can be smarter here, but for now yes. You immediately see the results in the frontend, which is the improvement

freddyaboulton commented:
What if we push developers towards using the js parameter instead of passing it in as the backend function? And the `js` decorator can instead be a function (and it doesn't even have to live in gradio so we can experiment faster).

So

```python
def update_ui():
    ...

btn.click(inputs=[...], outputs=[...], js=js(update_ui))
```

My feeling is that we should use the `js` parameter for this (and improve the DX of that where it falls short) rather than creating a new code path.

abidlabs commented:
> What if we push developers towards using the js parameter instead of passing it in as the backend function? And the js decorator can instead be a function (and it doesn't even have to live in gradio so we can experiment faster).

Ah that's a cool idea! The `js` decorator is already also a function, so this should basically work I think

freddyaboulton commented:
> yeah was thinking the same, creating a separate library for this, although the DX is slightly worse imo

We can include it as a required dep of gradio so that developers don't have to install anything else to try it out.

abidlabs commented:
Working on a standalone library here: https://github.com/abidlabs/grompy

abidlabs commented:
Thanks @freddyaboulton I've modified this PR a bit based on your feedback:

* First of all, I extracted all of the logic for js transpilation to [a separate library](https://github.com/abidlabs/grompy). Will allow us to move a bit faster as you said. Feedback welcome on the library, including potentially a cooler name
* Unfortunately its not possible to do exactly the approach you suggested `js=js(...)` because as I mentioned earlier, the flow is a bit different -- we still want to run the function in the backend, but we do not want to show the function's ui update in the frontend, otherwise that would defeat the purpose of a nicer UI. That being said, the decorator approach I suggested earlier is not great either since it cannot easily be applied to many kind of functions such as class methods and lambda functions. So instead, now you can set `js=True` in an event listener to get this transpilation behavior.

abidlabs commented:
Also take a look at `demo/blocks_essay/run.py` for example, which feels so much nicer now esp. on Spaces

whitphx commented on gradio/__init__.py:
```diff
@@ -225,6 +225,7 @@
     "close_all",
     "deploy",
     "get_package_version",
+    "js",
```
nit: looks like this is actually not exported from this module?

> "js" is specified in __all__ but is not present in modulePylance[reportUnsupportedDunderAll](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportUnsupportedDunderAll)

abidlabs replied:
thanks fixed @whitphx

whitphx reviewed (commented):
Interesting feature!
Sorry if I missed something but why don't you use the existing libraries such as https://github.com/TranscryptOrg/Transcrypt?

abidlabs commented:
> Sorry if I missed something but why don't you use the existing libraries such as https://github.com/TranscryptOrg/Transcrypt or https://github.com/flexxui/pscript?

Great question @whitphx. I thought about using existing libraries quite a bit but they all make different assumptions and provide a different developer experience that what would be ideal for Gradio users:

For Gradio developers, the most important thing is that the frontend function behaves exactly the same as the python function for any given inputs and outputs. If there's any ambiguity in how a function should be transpiled to js, or if it cannot be transpiled because e.g. it requires 3rd party libraries, an exception should be raised that explains how to fix the function so that it cannot be transpiled. Both parts of this are really important:

* **A function should never get transpiled incorrectly.** So for example, a function like:

```
import pscript

def calculator(num1, num2):
    return num1 + num2

print(pscript.py2js(calculator))
```

should not be transpiled, since depending on what `num1` and `num2` are (str/ints/arrays/sets/custom classes), the "+" operation can mean different things. Here, `pscript` transpiles this as:

```js
var _pyfunc_op_add = function (a, b) { // nargs: 2
    if (Array.isArray(a) && Array.isArray(b)) {
        return a.concat(b);
    } return a + b;
};
var calculator;
calculator = function flx_calculator (num1, num2) {
    return _pyfunc_op_add(num1, num2);
};
```

whereas `grompy` complains:

```
* Line 2: Ambiguous operation: Cannot determine types for 'num1 + num2'. Operation behavior may differ in JavaScript based on types.
>> return num1 + num2
```

* **Developers should be guided on how to modify their code so that it can be transpiled**: Unlike some other scenarios, Gradio developers are the ones writing their own Python functions, so they can make modifications as needed to allow transpilation to succeed. E.g. the solution to the problem above is to add type hints. 

If I update the function like this:

```py
import pscript
import grompy

def calculator(num1: int, num2: int):
    return num1 + num2

print(grompy.transpile(calculator))
```

Grompy complies:

```js
function calculator(num1, num2) {
    return (num1 + num2);
}
```

Pscript also complies but the returned function is unnecessarily complex since it doesn't use the type hints:

```js
var _pyfunc_op_add = function (a, b) { // nargs: 2
    if (Array.isArray(a) && Array.isArray(b)) {
        return a.concat(b);
    } return a + b;
};
var calculator;
calculator = function flx_calculator (num1, num2) {
    return _pyfunc_op_add(num1, num2);
};
```

Because I wanted to avoid false positives, and provide clear guidance to developers, I decided to create a new library that could allow us to control this experience. Of course, the natural question is that this is a time-consuming process and whether its worth investing in -- however, the `grompy` library was written almost entirely using o3-mini and is structured (single file, separate methods) in a way that makes it very easy for LLMs (or humans) to come in and make iterative improvements, so it has not required too much time investment so far.

abidlabs commented:
Oh finally I should mention that the last reason for creating a separate library is that there is some Gradio-specific transpilation logic, e.g. to handle component updates.

pngwn commented:
So i'm generally against this feature, I think it is pretty much impossible to either transpile correctly 100% of the time, or catch 100% of illegal transpilations. From other conversations we have had and the comments above, I'm assuming this is mostly LLM generated code and i don't agree with bringing such code into gradio. I've written a number of compilers and transpilers but I wouldn't call myself an expert, I wouldn't be comfortable hand-writing a compiler of this nature and bringing it into core, I certainly wouldn't be comfortable with predominantly LLM generated code. 

This is the main point for me, i do not believe that the code generated by large language models, that is not subsequently reviewed by a domain expert can be trusted. Especially not in a domain that is notoriously difficult/ borderline impossible to get right.

The second point is that I've seen this attempted a bunch of times and I've never seen it work, and the maintainers _were_ domain experts. The exception is for highly expressive statically typed languages with a fully sound type system, that typically gives you enough context to transpile most things. I will state, however, that even in these cases not everything can be translated.

> A function should never get transpiled incorrectly.

How can you guarantee this when both languages are incredible loose and flexible as well as having very different semantics?

> Developers should be guided on how to modify their code so that it can be transpiled

How are we expected to do this for all cases? That seems impossible and / or unmaintainable.

What if a user has written something that is semantically identical but syntactically different but our transpiler is not sophisticated enough to tell the difference? Being told to rewrite your code into the exact same thing is very frustrating imo.

dawoodkhan82 commented:
I think this could make for a cool opt-in optimization feature. Although like @pngwn mentioned, I'm not sure how many cases can be transpiled correctly.

What about a more generic developer facing api like an `optimize_with_ai` flag in `launch()`. Where internally we can experiment with passing the python function to convert to js, or wasm, or just optimizing the python function itself using an LLM. The user would know this is an experimental AI feature, and we have the flexibility of changing the implementation as needed. Not sure how realistic this is, but I think its time to play around with AI features in gradio core.

pngwn commented:
As an example of something i think should work and probably can't, maybe ever:

- pandas dataframe operations, the pandas API is vast and complex. In order support even a subset, we would essentially need to build pandas.js.
- in fact any time someone is using a 'type' that preprocesses the data to a python class, i would expect these to work because they are first class citizens on gradio. This means various media classes which are basically impossible to transpile with out a significant investment into developing a js runtime for them.

abidlabs commented:
I understand the concerns that you've raised @pngwn. Transpiling complex functions is not always going to be possible. However, when you're building a full Gradio app, you often have very simple functions (e.g. setting a button to be interactive or non-interactive based on an if/else condition) that are easy to implement in Python but the experience of using such a Gradio app is very frustrating because every UI interaction requires a complete server-side loop. This is what `js=True` is designed to address. Not particularly complex functions or functions that make use of 3rd party python libraries.

> So i'm generally against this feature, I think it is pretty much impossible to either transpile correctly 100% of the time, or catch 100% of illegal transpilations. From other conversations we have had and the comments above, I'm assuming this is mostly LLM generated code and i don't agree with bringing such code into gradio. I've written a number of compilers and transpilers but I wouldn't call myself an expert, I wouldn't be comfortable hand-writing a compiler of this nature and bringing it into core, I certainly wouldn't be comfortable with predominantly LLM generated code.

The code actually is quite straightforward, you can take a look at it here: https://github.com/abidlabs/grompy/blob/main/grompy/transpiler.py

> The exception is for highly expressive statically typed languages with a fully sound type system, that typically gives you enough context to transpile most things. I will state, however, that even in these cases not everything can be translated.

Right, so grompy takes a very different approach than most other transpilers. It (basically) only transpiles when the full context to transpile the code is in the function itself and all input parameters of that function are type-hinted. I give some examples of this in the readme. It also only works with a small subset of Python's standard library and _intentionally so_. Perhaps I need to do a better job of expectation-setting, but basically a subset of Python's standard library is whitelisted and everything else, Grompy will refuse to transpile. I can tell you though, that this is already quite useful in Gradio apps that I've built and in many of our example Gradio demos (I'll go through and add `js=True` to more to show where this can be useful).

pngwn commented:
I think this all aligns with my original thoughts and doesn't really change my position. 

I don't really see stuff like pandas or pillow as 'third party' because they are actually what we output the arguments as and are dependencies of gradio, you don't actually need to import anything to use them, they are already there.

abidlabs commented:
Even though there are several restrictions to use this right now (the python function has to be "simple", preprocess=False, postprocess=False), I've updated several demos to use `js=True` to showcase where it may be useful:

* `demo/blocks_essay/run.py`
* `demo/blocks_plug/run.py`
* `demo/blocks_random_slider/run.py`

yvrjsharma commented:
I absolutely loved this feature! It allowed me to create a chatbot input textbox with a dynamic `placeholder` value. It’s so exciting to add this kind of UI effect, which wasn’t possible before!

https://github.com/user-attachments/assets/f7e10096-7e5d-4baa-934a-2bf8f0bbeee2

It would be wonderful to retain this feature along with the suggested messaging: "UI update functions can now run on the client side" or something similar.

abidlabs commented:
As discussed in sync, I've modified the scope of this PR significantly. Now, we will transpile functions that *only* update the properties of components and do not take any inputs, e.g. functions that look like this:

```py
def update():
  return gr.Textbox(lines=4)
```

or 
```py
lambda : [gr.Textbox(lines=4), gr.Textbox(lines=4)]
```

This is not super useful but there are some cases where it can still improve the UX. For example, see the `demo/todo_list_js/run.py`

Recording on Spaces (original) with a relatively fast connection and a single user:

https://github.com/user-attachments/assets/44176d8b-e156-4505-8b3f-bf0a3cd12616

(This PR with `js=True`):


https://github.com/user-attachments/assets/c170eed9-08da-4717-a7e1-08851132230e



Still need to add validation & tests (will open for review after that) but the main pieces are in here.

abidlabs commented:
This PR should finally be ready for review! I've updated the parent comment to reflect the latest status of this PR. Also renamed the transpilation library from `grompy` --> `groovy`

yvrjsharma commented:
Working really nice for me 🔥 
With-js: https://huggingface.co/spaces/ysharma/update-components-with-js
Without-js: https://huggingface.co/spaces/ysharma/update-components-without-js

I can't update the component value this way, and I believe that's expected. Perhaps we could rephrase it to say "update component properties excluding `value`" or something similar.

abidlabs commented:
Good call yes @yvrjsharma!

freddyaboulton commented on gradio/blocks.py:
```diff
@@ -853,6 +859,11 @@ def set_event_trigger(
 
         rendered_in = LocalContext.renderable.get()
 
+        if js is True and inputs:
```
I wonder if we can selectively remove this restriction for chatinterface and interface's internal ui update functions and make them run client side super fast

freddyaboulton replied:
We can do that in a follow up though

abidlabs replied:
Yeah I definitely want to use `js=True` in `gr.ChatInterface` and `gr.Interface` wherever possible, but I'll do it in a follow-up PR in a comprehensive way so we don't have to make any sort of exception

abidlabs commented:
Thanks @freddyaboulton! Agreed on these suggestions, will implement

abidlabs commented:
> In @yvrjsharma 's demo, if you reset after adding an info to the component, null gets displayed. That's None -> null but I think we should filter out those values like we do in the backend

Actually I don't think we filter these out. Because we need to send `null` to the frontend to indicate "no info". But "nulls" are not being handled correctly nice catch

abidlabs commented:
Everything else should be handled here: https://github.com/abidlabs/groovy-transpiler/pull/1
